app/common/ts_util.c

```diff
@@ -0,0 +1,252 @@
+    ierr = PetscViewerASCIIPopTab(viewer);CHKERRQ(ierr);
+  }
+  ierr = DMGetDMTS(ts->dm,&sdm);CHKERRQ(ierr);
+  ierr = DMTSView(sdm,viewer);CHKERRQ(ierr);
+
+  ierr = PetscViewerASCIIPushTab(viewer);CHKERRQ(ierr);
+  ierr = PetscObjectTypeCompare((PetscObject)ts,TSSUNDIALS,&isundials);CHKERRQ(ierr);
+  ierr = PetscViewerASCIIPopTab(viewer);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode _TSLoad(TS ts, PetscViewer viewer)
+{
+  PetscErrorCode ierr;
+  PetscBool      isbinary;
+  PetscInt       classid;
+  char           type[256];
+  DMTS           sdm;
+  DM             dm;
+
+  PetscFunctionBegin;
+  PetscValidHeaderSpecific(ts,TS_CLASSID,1);
+  PetscValidHeaderSpecific(viewer,PETSC_VIEWER_CLASSID,2);
+  ierr = PetscObjectTypeCompare((PetscObject)viewer,PETSCVIEWERBINARY,&isbinary);CHKERRQ(ierr);
+  if (!isbinary) SETERRQ(PETSC_COMM_SELF,PETSC_ERR_ARG_WRONG,"Invalid viewer; open viewer with PetscViewerBinaryOpen()");
+
+  ierr = PetscViewerBinaryRead(viewer,&classid,1,NULL,PETSC_INT);CHKERRQ(ierr);
+  if (classid != TS_FILE_CLASSID) SETERRQ(PetscObjectComm((PetscObject)ts),PETSC_ERR_ARG_WRONG,"Not TS next in file");
+  ierr = PetscViewerBinaryRead(viewer,type,256,NULL,PETSC_CHAR);CHKERRQ(ierr);
+
+   /* missing content */
+   ierr = PetscViewerBinaryRead(viewer,&ts->ptime,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+   ierr = PetscViewerBinaryRead(viewer,&ts->steps,1,NULL,PETSC_INT);CHKERRQ(ierr);
+   ierr = PetscViewerBinaryRead(viewer,&ts->time_step,1,NULL,PETSC_DOUBLE);CHKERRQ(ierr);
+
+  ierr = TSSetType(ts, type);CHKERRQ(ierr);
+  if (ts->ops->load) {
+    ierr = (*ts->ops->load)(ts,viewer);CHKERRQ(ierr);
+  }
+  /* PETSC BUG FIX START */
+  if (!ts->adapt) {
+    ierr = TSAdaptCreate(PetscObjectComm((PetscObject)ts),&ts->adapt);CHKERRQ(ierr);
+    ierr = TSAdaptLoad(ts->adapt,viewer);CHKERRQ(ierr);
+  }
+  /* PETSC BUG FIX END */
+  ierr = DMCreate(PetscObjectComm((PetscObject)ts),&dm);CHKERRQ(ierr);
+  ierr = DMLoad(dm,viewer);CHKERRQ(ierr);
+  ierr = TSSetDM(ts,dm);CHKERRQ(ierr);
+  ierr = DMCreateGlobalVector(ts->dm,&ts->vec_sol);CHKERRQ(ierr);
+  ierr = VecLoad(ts->vec_sol,viewer);CHKERRQ(ierr);
+  ierr = DMGetDMTS(ts->dm,&sdm);CHKERRQ(ierr);
+  ierr = DMTSLoad(sdm,viewer);CHKERRQ(ierr);
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode tandem_TSView(TS ts)
+{ 
+  const char filename[] = "ts_checkpoint.bin";
+  PetscErrorCode ierr;
+  PetscViewer viewer;
+
+  ierr = PetscViewerBinaryOpen(PetscObjectComm((PetscObject)ts),filename,FILE_MODE_WRITE,&viewer);CHKERRQ(ierr);
+  ierr = _TSView(ts,viewer);CHKERRQ(ierr);
+  ierr = PetscViewerDestroy(&viewer);CHKERRQ(ierr);
+  /*
+  TSView(ts,PETSC_VIEWER_STDOUT_WORLD);
+  {
+    PetscReal time;
+    PetscInt tk;
+    TSGetStepNumber(ts,&tk); printf("  step %d\n",(int)tk);
+    TSGetTime(ts,&time); printf("  time %lf\n",time);
+  } 
+  */
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode tandem_TSLoad(TS ts)
+{
+  const char filename[] = "ts_checkpoint.bin";
```

Comment:
Please make the filename a function argument, then we can add it to the toml config.

Reply:
Will do.

Reply:
Done.